plugins/pipette/src/pipette.cc

```diff
@@ -0,0 +1,223 @@
+{
+}
+
+PipetteHandler::Ptr PipetteHandler::create()
+{
+  return Ptr(new PipetteHandler());
+}
+
+void PipetteHandler::computeValues(ViewInterface::Ptr view)
+{
+  gdk_threads_enter();
+  view->setStatusMessage("Computing colors values...");
+  gdk_threads_leave();
+
+  // Get the image rectangle
+  auto presentation = boost::static_pointer_cast<PresentationInterface>(view->getCurrentPresentation());
+  if(presentation == nullptr)
+  {
+    printf("PANIC: Current presentation does not implement PresentationInterface!\n");
+    gdk_threads_enter();
+    view->setStatusMessage("Error when requesting the image data.");
+    gdk_threads_leave();
+    enabled = true;
+    return;
+  }
+  auto image = presentation->getRect().toIntRectangle();
+
+  // Get the selection rectangle
+  int sel_x = std::min(selection->start.x, selection->end.x);
+  int sel_y = std::min(selection->start.y, selection->end.y);
+  auto sel_rect = Scroom::Utils::Rectangle<int>(sel_x, sel_y, selection->width(), selection->height());
+
+  // Intersect both rectangles to get the part of the selection that overlaps the image
+  auto rect = sel_rect.intersection(image);
+
+  // Get the average color within the rectangle
+  auto pipette = boost::dynamic_pointer_cast<PipetteViewInterface>(presentation);
+  if(pipette == nullptr)
+  {
+    printf("PANIC: Presentation does not implement PipetteViewInterface!\n");
+    gdk_threads_enter();
+    view->setStatusMessage("Error when requesting the image data.");
+    gdk_threads_leave();
+    enabled = true;
+    return;
+  }
+  auto colors = pipette->getPixelAverages(rect);
+
+  // If selection became null the plugin was switched off so ignore the result
+  enabled = true;
+  if(selection == nullptr)
+  {
+    return;
+  }
+
+  std::stringstream info;
+  info.precision(2);
+  fixed(info);
+
+  info << "Top-left: " << rect.getTopLeft();
+  info << ", Bottom-right: " << rect.getBottomRight();
+  info << ", Height: " << rect.getHeight();
+  info << ", Width: " << rect.getWidth();
+  if(!colors.empty()){
+    info << ", Colors:";
+    for(auto element : colors)
+    {
+      info << ' ' << element.first << ": " << element.second;
+    }
+  }
+
+  gdk_threads_enter();
+  view->setStatusMessage(info.str());
+  gdk_threads_leave();
+}
+
+////////////////////////////////////////////////////////////////////////
+// SelectionListener
+////////////////////////////////////////////////////////////////////////
+
+void PipetteHandler::onSelectionStart(GdkPoint, ViewInterface::Ptr)
+{
+}
+
+void PipetteHandler::onSelectionUpdate(Selection::Ptr s, ViewInterface::Ptr view)
+{
+  UNUSED(view);
+  if(enabled)
+  {
+    selection = s;
+  }
+}
+
+void PipetteHandler::onSelectionEnd(Selection::Ptr s, ViewInterface::Ptr view)
+{
+  if(enabled)
+  {
+    selection = s;
+
+    // Prevent more than one job
+    enabled = false;
+    Sequentially()->schedule(boost::bind(&PipetteHandler::computeValues, shared_from_this<PipetteHandler>(), view), pendingJobs);
```

Comment:
You should definitely improve the handling of multiple threads

Reply:
I have a debug build, were computing pixel values over an entire image takes 14 seconds. If I were to test on my netbook, of which I spoke earlier, it would probably be much more.

So if I selected the entire image, switched to measure, switched back to pipette, and then selected another image, I would have multiple background threads running. (well, except that the `Sequentially()` threadpool only has one thread).

Let's say that I were to select an entire image, and then select measure plugin, and go do some measurements. By the time the background thread finishes, it will re-enable Pipette plugin too, and now I have two subscribers for selection events.

The list goes on. I'm a bit confused at how you can so boldly claim that there are no race conditions, when the consensus in the industry seems to be that multithreading is hard and should be avoided. A consensus that you guys seem to agree with, given that you want to use only one thread for this endeavor, when you could trivially speed it up by using more.

To wrap up
* Thread handling, as it currently is, is broken and should be rethought. No amount of mutexes will fix that.
* Mutexes are not only for protecting something from concurrent access (though I hope I've proven that that, too, will happen), but also keeps the compiler and the processor from reordering instructions, and adds cpu-cache-synchronisation instructions. In short, if you want to make any prediction about what your multithreaded program will do (and be correct), then you'd better use them.

Reply:
Yes you are correct, I realized this same edge case over dinner as well. Though I do not really call issues like this a race condition (even though maybe I should?) since they are almost completely deterministic. I for some reason decided that the `enabled` flag was not accessible from the UI thread anymore after becoming `false` but that is not true. 

So I'll fix this, though I do not see how a mutex would help with this. The general idea is to disable the handling of events when a background job is in progress and there should only ever be one background job. The `onSelectionEnd` method is executed on the UI thread and therefore as far as I know there will never be called by multiple threads executing this method at the same time. Therefore setting a dedicated flag that prevents `onSelectionEnd` from executing inside `onSelectionEnd` should prevent a second job from being submitted. Then setting unsetting this flag at the end of the background job makes it so a new job can be submitted, but at this point the only background job has also finished executing. While a mutex could be used to prevent a second job from being submitted as far as I know the UI thread would in this case lock up to wait for the mutex lock. Where/how should a mutex be used here? As far as I can tell a boolean that is not accessible to the UI thread should suffice here.

> I have a debug build, were computing pixel values over an entire image takes 14 seconds. If I were to test on my netbook, of which I spoke earlier, it would probably be much more.

What kind of image are you testing this on? I'm personally developing on a laptop with a CPU that was released 10 years ago and even then I can still compute the average color value of an image that is 14391 x 6232 pixels (URD has 10k x 10k) within about 3 seconds (URD has 15sec).

Reply:
Two things

**Problems solved with mutices**
When you have a variable that is accessed from multiple threads (in your implementation, that's `jobRunning` and `selection`), then you need to use some synchronization primitive (i.e. mutices) to safeguard them. I always thought that was commonly understood, and I did not expect to have to argue the point. Do you really need me to elaborate?

**Problems not solved with mutices**
There's still the scenario that if I disable and enable Pipette, and then make a selection, all while the previous computeValues is still running, then my new selection will be ignored. I don't think that's ok.

**Your other questions**
I'm testing on the Onyx evaluation image provided by canon. (20704x11660 pixels). More importantly, I've turned off optimizations. Optimizations do a lot to speed up scroom.

Note though, that speed is not my point. It is correctness. If I chose to run scroom on a raspberry pi 1, it should still run correctly (albeit possibly unbearably slow, and maybe with some memory constraints).

However, by artificially slowing down scroom, I can now manually trigger bugs in your implementation that would be much harder to trigger (though still present) at "normal" speed. And turning off optimization is an easier way to accomplish that goal, than hooking up my raspberry pi :wink:

Reply:
So what you're actually referring to is a data race instead of a race condition which is fair and I've updated the implementation to handle this. Apparently data races also appear to have more serious side effects in C++ compared to other languages?

The case where a background job remains running is fair (and known), however this is not something we have enough time left for to handle.